### PR TITLE
Improve error message on invalid values to Distribution methods

### LINKED ISSUE #18133: Descriptive error message for distribution constraint violation
## 🚀 Feature

The current error message raised by constraint violations in distribution constructors (when `validate_args=True`) is a generic message that doesn't say what the problem is, e.g.:
```
<...>/torch/distributions/distribution.py", line 36, in __init__
    raise ValueError("The parameter {} has invalid values".format(param))
ValueError: The parameter probs has invalid values
```

## Motivation

When I see this error, I have to identify the source of the problematic parameter, add logging code, run the program again (which might take a long time), and hope that it fails again in the same place (which it might not due to stochasticity).

## Pitch

The particular error I encountered was for the `Categorical` distribution, where I had a long probability vector (4096 entries) that caused the sum of the normalized probabilities to fall outside the error tolerance in the constraint check. The desired message would be something like:
```
ValueError: The parameter probs has invalid values: |sum(probs) - 1| > 1e-6
```
which describes what the problem actually is. Note that in this case it was the `_Simplex` constraint that failed, but it can fail for two different reasons (negative entries, or normalized sum not close enough to 1).

On a related note, perhaps the error tolerance could account for the number of entries in `probs` to avoid raising a spurious error.

### PR DESCRIPTION
**Pull request by fritzo**
Fixes #18133

### CODE REVIEW

**Comment by fritzo** [comment]
Any ideas how to avoid these nonsensical failures in mypy?
```
torch/distributions/negative_binomial.py:37:9: error: Property "_param" defined in "NegativeBinomial" is read-only  [misc]
torch/distributions/categorical.py:61:9: error: Property "_param" defined in "Categorical" is read-only  [misc]
torch/distributions/categorical.py:67:16: error: "Categorical" has no attribute "_dict__"  [attr-defined]
torch/distributions/binomial.py:49:9: error: Property "_param" defined in "Binomial" is read-only  [misc]
torch/distributions/continuous_bernoulli.py:57:9: error: Property "_param" defined in "ContinuousBernoulli" is read-only  [misc]
torch/distributions/relaxed_bernoulli.py:43:9: error: Property "_param" defined in "LogitRelaxedBernoulli" is read-only  [misc]
```

File: .github/workflows/periodic-pytorch-win-vs2019-cuda11-cudnn8-py3.yml [hunk 1]
```diff
@@ -80,34 +80,34 @@ jobs:
     if: ${{ github.repository_owner == 'pytorch' }}
     runs-on: ubuntu-18.04
     env:
+      TEST_RUNNER_TYPE: windows.8xlarge.nvidia.gpu
```
Comment (fritzo) [inline .github/workflows/periodic-pytorch-win-vs2019-cuda11-cudnn8-py3.yml:83]: This seems weird, but a test failure told me to run `.github/scripts/generate_ci_workflows.py` and push :confused:

File: torch/distributions/distribution.py [hunk 2]
```diff
@@ -273,8 +273,15 @@ def _validate_sample(self, value):
                           '`validate_args=False` to turn off validation.')
             return
         assert support is not None
-        if not support.check(value).all():
-            raise ValueError('The value argument must be within the support')
+        valid = support.check(value)
+        if not valid.all():
+            invalid_value = value.data[~valid]
+            error = ValueError(
```
Comment (neerajprad) [inline torch/distributions/distribution.py:279]: Do you think we should also specify the support in the error message for arguments checking in `__init__`?

**Review by neerajprad — COMMENTED** [review]
Thank you for improving the error message! Could you remove the change to the CI workflow? I think we can handle that in a separate PR if needed and get this merged sooner.

File: test/distributions/test_distributions.py [hunk 3]
```diff
@@ -4555,7 +4555,10 @@ def test_invalid(self):
         for Dist, params in BAD_EXAMPLES:
             for i, param in enumerate(params):
                 try:
-                    with self.assertRaises(ValueError):
+                    with self.assertRaisesRegex(
```
Comment (neerajprad) [inline test/distributions/test_distributions.py:4558]: Just seeing the test errors now -- the error message for arguments checking is different currently and only says that the parameter has incorrect value without specifying the support. So this regex match fails.

Comment (fritzo) [reply test/distributions/test_distributions.py:4558]: oops, looks like I updated the wrong test; I'll fix shortly...

File: .github/workflows/periodic-pytorch-win-vs2019-cuda11-cudnn8-py3.yml (hunk 1 above)
Comment (fritzo) [reply .github/workflows/periodic-pytorch-win-vs2019-cuda11-cudnn8-py3.yml]: ok, reverted.

File: torch/distributions/distribution.py (hunk 2 above)
Comment (fritzo) [reply torch/distributions/distribution.py]: done

**Comment by fritzo** [comment]
I believe remaining test errors are unrelated.

**Review by neerajprad — APPROVED** [review]

**Comment by fritzo** [comment]
Thanks for reviewing @neerajprad!

**Comment by eellison** [comment]
cc @neerajprad looks like this broke lint

**Comment by neerajprad** [comment]
> cc @neerajprad looks like this broke lint

@eellison: As discussed in another thread, this doesn't touch any files that are currently being flagged by mypy. I think the change is due to changes in the linter settings itself introduced in https://github.com/pytorch/pytorch/pull/61119.

**Pull request closed without merging**